plugins/genesys/fu-genesys-usbhub-device.c

```diff
@@ -835,84 +920,83 @@ fu_genesys_usbhub_device_setup(FuDevice *device, GError **error)
 		self->flash_sector_size = sector_size;
 
 	/* setup firmware parameters */
-	switch (self->isp_model) {
-	case ISP_MODEL_HUB_GL3523:
+	switch (self->chip.model) {
+	case ISP_MODEL_HUB_GL3523: {
+		gint var = 0;
+		guint8 bonding = 0;
 		self->fw_bank_addr[0] = 0x0000;
 		self->fw_bank_addr[1] = 0x8000;
 		self->extend_size = GL3523_PUBLIC_KEY_LEN + GL3523_SIG_LEN;
-		if (self->isp_revision == 50) {
+
+		if (self->chip.revision == 50) {
 			self->fw_data_total_count = 0x8000;
-			if (!fu_genesys_usbhub_device_get_fw_size(self, 0, error))
-				return FALSE;
+			self->fw_key_addr[0] = 0x7D00;
+			self->fw_key_addr[1] = 0xFD00;
 		} else {
 			self->fw_data_total_count = 0x6000;
-			self->code_size = self->fw_data_total_count;
+			self->fw_key_addr[0] = 0x6000;
+			self->fw_key_addr[1] = 0xE000;
+			self->code_size[0] = self->code_size[1] = self->fw_data_total_count;
 		}
-		break;
-	case ISP_MODEL_HUB_GL3590:
-		if (!fu_genesys_usbhub_device_get_fw_size(self, 0, error))
+
+		var = fu_genesys_tsdigit_value((gchar)self->dynamic_ts.bonding);
+		if (var == -1) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "GL3523 bonding value(%hhX) is wrong",
+				    self->dynamic_ts.bonding);
 			return FALSE;
+		}
+		bonding = (var & 0xff);
+
+		if (self->static_ts.tool_string_version < TOOL_STRING_VERSION_BONDING_QC)
+			bonding <<= 1;
+		if (self->dynamic_ts.running_mode == 'M') {
+			self->running_bank = BANK_MASK_CODE;
+		} else if ((bonding & 0x10) > 0) {
+			self->running_bank = BANK_SECOND;
+		} else {
+			self->running_bank = BANK_FIRST;
+		}
+		fu_device_add_instance_u8(device, "BONDING", (bonding & GL3523_BONDING_VALID_BIT));
+		break;
+	}
+	case ISP_MODEL_HUB_GL3590: {
+		guint8 bonding = self->dynamic_ts.bonding;
```

Comment:
This temporary variable does not look to be necessary.

BTW, is it normal the process to get the bonding value is different between GL3590 and GL3523?

Reply:
Sorry, I'm not sure your question. Do you mean the process to get right bonding value from GL3590 and GL3523 has different?

As FuGenesysDynamicToolString comments, the `bonding` byte contains many information for GL3590 and GL3523. For new ICs(e.g. GL3525) they have new dynamic tool string format.

Reply:
I mean that for GL3523, you do

``` c
var = fu_genesys_tsdigit_value(...);
var = (var & 0xff)
if (self->static_ts.tool_string_version < TOOL_STRING_VERSION_BONDING_QC)
    var <<= 1;
self->bonding = (var & GL3523_BONDING_VALID_BIT);  
```

while it is that short for GL3590

``` c
self->bonding = (self->dynamic_ts.bonding & GL3590_BONDING_VALID_BIT);
```

Reply:
which means that, the GL3523 is using the Static TS, and the GL3590 is using the Dynamic TS.

I am just checking if that is correct.

Reply:
> I am just checking if that is correct.

The spec history says the improvement in newer tool string version; the bonding byte has shift to indicate QC supporting for GL3523.

How about add `fu_genesys_usbhub_device_get_bonding(FuGenesysUsbhubDevice *self, GError **error)` to do that?

Reply:
I was just asking.

If you think this makes the code nicer go ahead. I have no strong opinion for now.